src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.Tasks/JoinVerticals.cs

```diff
@@ -0,0 +1,290 @@
+    private const string _artifactNameSuffix = "_Artifacts";
+    private const string _assetsFolderName = "assets";
+    private const string _packagesFolderName = "packages";
+    private const int _retryCount = 10;
+
+    public override bool Execute()
+    {
+        ExecuteAsync().Wait();
+        return !Log.HasLoggedErrors;
+    }
+
+    private async Task ExecuteAsync()
+    {
+        List<XDocument> verticalManifests = VerticalManifest.Select(xmlPath => XDocument.Load(xmlPath.ItemSpec)).ToList();
+
+        if (!string.IsNullOrEmpty(VerticalSubSet))
+        {
+            var verticalSubSet = VerticalSubSet.Split(';');
+            verticalManifests = verticalManifests.Where(manifest => verticalSubSet.Contains(GetRequiredRootAttribute(manifest, _verticalNameAttribute))).ToList();
+        }
+
+        // Find the main manifest, and remove it from the list
+        XDocument mainVerticalManifest = verticalManifests.FirstOrDefault(manifest => GetRequiredRootAttribute(manifest, _verticalNameAttribute) == MainVertical)
+            ?? throw new ArgumentException($"Couldn't find main vertical manifest {MainVertical} in vertical manifest list");
+
+        int removed = verticalManifests.RemoveAll(manifest => GetRequiredRootAttribute(manifest, _verticalNameAttribute) == MainVertical);
+
+        if (removed > 1)
+        {
+            throw new ArgumentException($"Found more than one main vertical manifest {MainVertical} in vertical manifest list");
+        }
+
+        if (!Directory.Exists(MainVerticalArtifactsFolder))
+        {
+            throw new ArgumentException($"Main vertical artifacts directory {MainVerticalArtifactsFolder} not found.");
+        }
+
+        string mainVerticalName = GetRequiredRootAttribute(mainVerticalManifest, _verticalNameAttribute);
+
+        Dictionary<string, XElement> packageElements = [];
+        Dictionary<string, XElement> blobElements = [];
+
+        List<string> addedPackageIds = AddMissingElements(packageElements, mainVerticalManifest, _packageElementName);
+        List<string> addedBlobIds = AddMissingElements(blobElements, mainVerticalManifest, _blobElementName);
+
+        CopyMainVerticalAssets(Path.Combine(MainVerticalArtifactsFolder, _packagesFolderName), Path.Combine(OutputFolder, _packagesFolderName));
+        CopyMainVerticalAssets(Path.Combine(MainVerticalArtifactsFolder, _assetsFolderName), Path.Combine(OutputFolder, _assetsFolderName));
+
+        foreach (XDocument verticalManifest in verticalManifests)
+        {
+            string verticalName = GetRequiredRootAttribute(verticalManifest, _verticalNameAttribute);
+
+            addedPackageIds = AddMissingElements(packageElements, verticalManifest, _packageElementName);
+            addedBlobIds = AddMissingElements(blobElements, verticalManifest, _blobElementName);
+
+            if (addedPackageIds.Count > 0 || addedBlobIds.Count > 0)
+            {
+                await DownloadArtifactFiles(
```

Comment:
Suggest adding parallelism to this. You don't need to have the files downloaded until the task exits I think, so you could implement this by adding the tasks to a collection and awaiting on a WaitAll of all those tasks. You can then add a second level of parallelism in `DownloadArtifactFiles` by doing a similar thing over all files you want to download.

To avoid completely smashing the machine to bits, add a throttle mechanism in `DownloadFileFromArtifact`. You can see how a throttle is used here:

https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs#L1173-L1189

There are also a few other ways to do this (e.g. return a lists of tasks from `DownloadArtifactFiles`) if you so prefer.

Reply:
parallelization got the time down to 5 mins from about 30

Reply:
💃

Reply:
Awesome. Great win.